File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

/**
# Boj 15686 치킨거리
## 문제이해
- N*N도시
- 빈칸,집,가게 = {0,1,2}
- 치킨거리: 현재 집에서 가장 가까운 가게 거리
- 도시치킨거리: 각집의 치킨거리의 합
- M은 살아남을 치킨집 개수

## 설계
1. 치킨집 고르는 경우의 수 구하기
- 치킨집 최대 13개 중에서 M개 고르기
- 최대 13C6 

2. 각 집의 치킨거리 구하기
- 살아남은 치킨 -> 집 이중 돌면서 최소 치킨거리 구하기
- 이 때 집마다 돌면서, 최소 치킨거리 갱신
- 13 * 13
3. 도시치킨거리 최소합 갱신
- 집마다 돌면서 sum 구하기
- sum의 min값 구하기

---
집의 개수는 2N을 넘지않음 => dist의 MX 102 설정;
치킨집과 M의 최대 개수는 13 => arr의 MX 20설정
*/


int n, m;

vector<pair<int, int>> stores;
vector<pair<int, int>> houses;
int dists[102];
vector<vector<int>> tests;
vector<int> arr;

void recur(int cur, int st){
    if(cur == m){
        tests.push_back(arr);
        return;
    }
    for(int i = st; i < stores.size(); i++){
        arr.push_back(i);
        recur(cur+1, i+1);
        arr.pop_back();
    }
}

int get_dist(auto& x, auto& y){
    return abs(x.first - y.first) + abs(x.second - y.second);
}

int main()
{
    cin >> n >> m;
    
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            int v;
            cin >> v;
            if(v == 2) stores.push_back({i,j});
            else if(v == 1) houses.push_back({i,j});
        }
    }
    
    recur(0, 0);

    int mn = 1e9;
    // 1. 살아남은 가게 경우의수
    for(auto t : tests){
        for(int i = 0; i < 102; i++){
            dists[i] = 1e9;
        }
        // 2-1. 먼저 살아남은 치킨집 돌고
        for(int alive_idx: t){
            auto s = stores[alive_idx];
             // 2-2. 집 돌면서 최솟값 갱신
            for(int i = 0; i < houses.size(); i++){
                auto h = houses[i];
                dists[i] = min(dists[i], get_dist(h, s));
            }
        }
        // 3. 도시거리 구해서, 최솟값 갱신
        int sum = 0;
        for(int i = 0; i < houses.size();i++){
            sum += dists[i];
        }
        
        mn = min(sum, mn);
    }
    
    cout << mn;

    return 0;
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static final int HOUSE = 1, STORE = 2;
    static int board[][] = new int[52][52];
    static int N, M;
    static ArrayList<Integer[]> stores = new ArrayList<>();
    static ArrayList<Integer[]> houses = new ArrayList<>();
    static int selected[] = new int[22];
    static int mn = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        for (int r = 0; r < N; r++) {
            st = new StringTokenizer(br.readLine());
            for (int c = 0; c < N; c++) {
                board[r][c] = Integer.parseInt(st.nextToken());
                // 집과 치킨집 리스트에 저장
                if (board[r][c] == HOUSE) {
                    houses.add(new Integer[]{r, c});
                } else if (board[r][c] == STORE) {
                    stores.add(new Integer[]{r, c});
                }
            }
        }

        recur(0, 0);

        System.out.println(mn);
    }

    static void recur(int cur, int st) {
        if (cur == M) {
            solution();
            return;
        }

        for (int i = st; i < stores.size(); i++) {
            selected[cur] = i;
            recur(cur + 1, i + 1);
        }
    }

    static void solution() {
        int cityDist = 0;
        for (Integer[] house : houses) {
            int chickenDist = Integer.MAX_VALUE;
            for (int i = 0; i < M; i++) {
                int idx = selected[i];
                Integer[] store = stores.get(idx);
                int dist = getDist(house[0], house[1], store[0], store[1]);
                chickenDist = Math.min(dist, chickenDist);
            }
            cityDist += chickenDist;
        }
        mn = Math.min(mn, cityDist);
    }

    static int getDist(int r1, int c1, int r2, int c2) {
        return Math.abs(r1 - r2) + Math.abs(c1 - c2);
    }
}